### PR TITLE
fix: add rotate and step additonal checks + use custom errors for gas…

### PR DESCRIPTION
**Pull request by nmlinaric**
This PR adds:
- check for rotate and step (if value already exists in mappings it will revert with Error)
- removes unused `head` const
- replaces `require` statements with custom errors to save on gas

### CODE REVIEW

File: src/Spectre.sol [hunk 1]
```diff
@@ -152,15 +145,9 @@ contract Spectre {
...
-            _publicInputsCommitment == publicInputsCommitment,
-            "Invalid public inputs commitment"
-        );
+        if (_publicInputsCommitment == publicInputsCommitment) revert InvalidPublicInputsCommitment();
 
-        require(
-            _syncCommitteePoseidon == syncCommitteePoseidon,
-            "Invalid sync committee poseidon"
-        );
+        if (_syncCommitteePoseidon == syncCommitteePoseidon) revert InvalidSyncComitee();
```
Comment (nulltea) [inline src/Spectre.sol]:
```suggestion
        if (_syncCommitteePoseidon != syncCommitteePoseidon) revert InvalidSyncComitee();
```

File: src/Spectre.sol [hunk 2]
```diff
@@ -152,15 +145,9 @@ contract Spectre {
             bytes32(proof[384 + 32:384 + 2 * 32])
         );
 
-        require(
-            _publicInputsCommitment == publicInputsCommitment,
-            "Invalid public inputs commitment"
-        );
+        if (_publicInputsCommitment == publicInputsCommitment) revert InvalidPublicInputsCommitment();
```
Comment (nulltea) [inline src/Spectre.sol]:
```suggestion
        if (_publicInputsCommitment != publicInputsCommitment) revert InvalidPublicInputsCommitment();
```

**Review by nulltea — CHANGES_REQUESTED** [review]

File: src/Spectre.sol [hunk 3]
```diff
@@ -93,36 +95,27 @@ contract Spectre {
...
-            "Sync committee not yet set for this period"
+        // *rotation phase*
+        // This proof checks that the given poseidon commitment and SSZ commitment to the sync committee are equivalent and that
+        // that there exists an SSZ proof that can verify this SSZ commitment to the committee is in the state
+        uint256 currentPeriod = _getSyncCommitteePeriod(
+            stepInput.finalizedSlot
         );
+        uint256 nextPeriod = currentPeriod + 1;
+        if (syncCommitteePoseidons[nextPeriod] != 0) revert SyncCommitteeAlreadySet();
+        if (syncCommitteePoseidons[attestingPeriod] != 0) revert SyncCommiteeNotYetSetForPeriod();
```
Comment (lastperson) [inline src/Spectre.sol]:
```suggestion
        if (syncCommitteePoseidons[attestingPeriod] == 0) revert SyncCommiteeNotYetSetForPeriod();
```

**Review by lastperson — CHANGES_REQUESTED** [review]
Shouldn't there be tests for of the reverts?

**Comment by nmlinaric** [comment]
> Shouldn't there be tests for of the reverts?

~I'll add them in the following days, wanted to confirm if all the checks logic is valid.~
There are some tests in Spectre repo written in rust, I'll extend those since the initial logic is already prepared.

**Review by nulltea — APPROVED** [review]

**Review by lastperson — APPROVED** [review]

**Pull request merged**